cpp-package/example/inference/sentiment_analysis_rnn.cpp

```diff
@@ -0,0 +1,495 @@
+ * Ensure to use the same dictionary file that was used for training the network.
+ */
+void Predictor::LoadDictionary(const std::string& input_dictionary) {
+  if (!FileExists(input_dictionary)) {
+    LG << "Dictionary file " << input_dictionary << " does not exist";
+    throw std::runtime_error("Dictionary file does not exist");
+  }
+  LG << "Loading the dictionary file.";
+  std::ifstream fi(input_dictionary.c_str());
+  if (!fi.is_open()) {
+    std::cerr << "Error opening dictionary file " << input_dictionary << std::endl;
+    assert(false);
+  }
+
+  std::string line;
+  std::string word;
+  int index;
+  while (std::getline(fi, line)) {
+    std::istringstream stringline(line);
+    stringline >> word >> index;
+    wordToIndex[word] = index;
+  }
+  fi.close();
+}
+
+
+/*
+ * The function returns the index associated with the word in the dictionary.
+ * If the word is not present, the index representing "<unk>" is returned.
+ * If the "<unk>" is not present then 0 is returned.
+ */
+float Predictor::GetIndexForWord(const std::string& word) {
+  if (wordToIndex.find(word) == wordToIndex.end()) {
+    if (wordToIndex.find("<unk>") == wordToIndex.end())
+      return 0;
+    else
+      return static_cast<float>(wordToIndex["<unk>"]);
+  }
+  return static_cast<float>(wordToIndex[word]);
+}
+
+/*
+ * The function populates the input vector with indices from the dictionary that
+ * correspond to the words in the input string.
+ * The function returns the number of words in the input line.
+ */
+int Predictor::ConvertToIndexVector(const std::string& input, std::vector<float> *input_vector) {
+  std::istringstream input_string(input);
+  input_vector->clear();
+  const char delimiter = ' ';
+  std::string token;
+  size_t words = 0;
+  while (std::getline(input_string, token, delimiter) && (words <= input_vector->size())) {
+    input_vector->push_back(GetIndexForWord(token));
+    words++;
+  }
+  return words;
+}
+
+
+/*
+ * The function returns the index at which the given symbol name will appear
+ * in the output vector of NDArrays obtained after running the forward pass on the executor.
+ */
+int Predictor::GetIndexForOutputSymbolName(const std::string& output_symbol_name) {
+  int index = 0;
+  for (const std::string op : net.ListOutputs()) {
+    if (op == output_symbol_name) {
+      return index;
+    } else {
+      index++;
+    }
+  }
+  throw std::runtime_error("The output symbol name can not be found");
+}
+
+
+/*
+ * The function finds the closest bucket for the given num_words in the input line.
+ * If the exact bucket key exists, function returns that bucket key.
+ * If the matching bucket key does not exist, function looks for the next bucket key
+ * that is greater than given num_words.
+ * If the next larger bucket does not exist, function looks for the previous bucket key
+ * that is lesser than given num_words.
+ */
+int Predictor::GetClosestBucketKey(int num_words) {
```

Comment:
it seems that you can simply use lower_bound and return the largest key if not found.

Reply:
If we use lower_bound, we will always find a bucket that is smaller than the input word length (unless the input word length is smaller than the smallest bucket size). In that case, we will trim the input line and may lose the information.  Therefore, I chose to find the next largest bucket so that we will be able to consider all the words in input sentence.

Reply:
> If we use lower_bound, we will always find a bucket that is smaller than the input word length

I'm not sure if this is true. From cpprefenrece on [std::map::lower_bound](https://en.cppreference.com/w/cpp/container/map/lower_bound):

> Returns an iterator pointing to the first element that is not less than key.

In your case the `key` is input word length.

Reply:
@szha I agree, with lower_bounds we will get matching or a smaller bucket key.  My concern is that, we will always find such bucket and we will trim the input sequence. In that process, we would lose some of the information. Will it be acceptable?

Reply:
Unless I read the document wrong, it won't get you smaller key. `lower_bound` returns the first element (i.e. bucket key) that's no less than (i.e. larger or equal to) the key (i.e. input sequence length).

Reply:
For input sequence that's larger than the largest key, clipping is necessary.

Reply:
@szha  Fixed the logic in the latest update. Getting the lower bound is sufficient and accurate.
Thanks!